### PR TITLE
feat(INF-1126): sweep retention range to completion

### PR DESCRIPTION
**Pull request by Byron-X-22**
## Summary
- make execute-mode single-block retention continue as new until a post-processing empty selection confirms the frozen sweep is complete
- freeze the destructive set with the dry-run `EligibilityCutoff` through selection, reconciliation, planning, apply-time revalidation, and every continuation
- pin the workflow-resolved effective block tag into every new-sweep activity request, including explicit tag zero, while preserving deployed legacy Temporal command inputs
- read pending and due cohorts in one read-only repeatable-read PostgreSQL snapshot and merge row-disjoint aggregates with exact counts
- preserve exact approval matching for direct admin callers and require an explicit workflow-only containing-envelope mode after independent cohort validation
- accept pre-sweep activity payloads without a cutoff using legacy wall-clock behavior, without weakening new sweep requests

## Verification
- `TEST_TYPE=unit go test -count=1 -p=1 ./...`
- `TEST_TYPE=unit go test -count=1 ./internal/storage/retirement ./internal/workflow`
- `TEST_TYPE=unit go test -count=1 -run SingleBlockRetention ./internal/workflow/activity`
- `TEST_TYPE=unit go test -count=1 ./internal/workflow/activity`
- `go vet -printfuncs=wrapf,statusf,warnf,infof,debugf,failf,equalf,containsf,fprintf,sprintf ./...`
- `errcheck -ignoretests -ignoregenerated ./...`
- `ineffassign ./...`
- `goimports` and `git diff --check`
- independent adversarial re-review: no constructive findings remain

Linear: https://linear.app/cipherowl/issue/INF-1126

### CODE REVIEW

File: internal/workflow/single_block_retention.go [hunk 1]
```diff
@@ -287,12 +337,30 @@ func (w *SingleBlockRetention) execute(
 				}
 			}
 		}
+		if request.Execute && rangeSweepEnabled && selected.HasMore {
```
Comment (watch-owl) [inline internal/workflow/single_block_retention.go]: This completion/continuation decision is based on `HasMore` from the select that ran *before* the current batch was processed. The due query uses `CURRENT_TIMESTAMP`, and processing can run for hours; a second cohort can become due (or a resumable cohort can appear) while this batch is running. If the initial snapshot contained exactly `Limit` cohorts, `HasMore` is false and line 355 reports `SweepCompleted=true` without proving the range is empty at completion. Please continue after every non-empty execute batch (so the next run performs the empty probe), or perform an equivalent post-processing reselect before declaring the sweep complete. Add a test where a non-empty batch has `HasMore=false` and completion requires a subsequent empty selection.

File: internal/workflow/single_block_retention.go [hunk 2]
```diff
@@ -287,12 +337,30 @@ func (w *SingleBlockRetention) execute(
...
+			nextRequest := *request
```
Comment (watch-owl) [inline internal/workflow/single_block_retention.go:395]: `nextRequest` keeps the caller's raw `Tag`, not the resolved `tag`. For the normal omitted-tag case (`Tag == 0`), each continuation recomputes the effective tag from the newly read config at line 198. If the stable tag changes between runs, checkpoint validation fails and strands a partially completed sweep; it does not preserve the original effective tag as the PR claims. Please resolve continuation runs from `Checkpoint.EffectiveTag` (including the real tag-zero/`MaxUint32` sentinel case), or otherwise pin the resolved tag in the continuation request, and cover a config stable-tag change between runs.

**Review by watch-owl — CHANGES_REQUESTED** [review]
## Code Review Summary

### Warnings

• The sweep decides it is complete from the selector snapshot taken before processing. A cohort can become due or pending while the selected batch is being processed, so the final result can report `sweep_completed=true` even though eligible work remains in the approved range.

• Continue-as-new validates the effective tag but does not pin it. When the caller leaves `Tag` at its normal default (`0`), a stable-tag config change between runs causes the next run to resolve a different tag and fail instead of continuing the approved sweep on the original effective tag.

### Verification

• `TEST_TYPE=unit go test -count=1 ./internal/workflow ./internal/workflow/activity ./internal/storage/retirement` passed locally.
• `ci/circleci: build_and_test` is green on the reviewed head.

File: internal/workflow/single_block_retention.go [hunk 3]
```diff
@@ -161,21 +215,34 @@ func (w *SingleBlockRetention) execute(
...
 		activityCtx := w.withActivityOptions(ctx)
 		selected, err := w.activity.Select(activityCtx, &activity.SingleBlockRetentionSelectRequest{
-			Tag:         tag,
-			StartHeight: request.StartHeight,
-			EndHeight:   request.EndHeight,
-			Limit:       maxObjectRanges,
+			// Preserve the caller's tag encoding until the activity resolves it
+			// exactly once. In particular, MaxUint32 is the only encoding for
+			// effective tag zero.
+			Tag:               request.Tag,
```
Comment (watch-owl) [inline internal/workflow/single_block_retention.go]: The continuation now pins `tag`, but this still passes the caller's raw value (normally `0`) to the activity. The workflow's `cfg` was captured through Temporal history, whereas `executeSelect` and `planRequest` each call `a.config.GetEffectiveBlockTag(request.Tag)` against the worker's live config. If the stable tag changes after this run's config SideEffect was recorded—e.g. the workflow resumes on a newly deployed worker—selection and processing can target the new tag while `result.Tag` and the checkpoint retain the old tag. Please pass `encodeSingleBlockRetentionEffectiveTag(tag)` to the initial select, completion probe, and process requests so every activity resolves to the workflow-pinned effective tag. Add a regression test that changes the worker's stable tag after the workflow-side resolution and verifies all activity requests retain the original effective tag.

**Review by watch-owl — CHANGES_REQUESTED** [review]
## Validation Review

The completion-snapshot finding is fixed: the eligibility cutoff freezes the approved destructive set, and a post-processing selector probe prevents false completion.

The effective-tag finding remains partially unresolved at the activity boundary. The workflow pins the tag for continuation, but the current run still passes the caller's raw tag encoding to each activity, where it is resolved again from live worker config. A worker/config change can therefore make selection or processing target a different tag than the workflow result and checkpoint record.

Focused unit packages pass locally. CircleCI is currently pending (and is not configured as a required status check).

File: internal/workflow/single_block_retention.go (hunk 3 above)
Comment (Byron-X-22) [reply internal/workflow/single_block_retention.go]: Addressed in dc29cdc. The new sweep version now passes encodeSingleBlockRetentionEffectiveTag(tag) to the initial Select, every Process call/retry, and the completion probe, so activity workers cannot re-resolve an omitted tag against newer live config. The DefaultVersion path still passes the historically resolved numeric tag to preserve deployed Temporal command compatibility. Regression coverage verifies default stable-tag pinning across Select/Process, explicit effective tag zero, and a legacy config change between Select and Process. Focused tests, full serial unit tests, full vet, errcheck, ineffassign, goimports, diff check, and an independent adversarial re-review pass.

**Review by watch-owl — APPROVED** [review]
## Validation Review

No blocking findings remain on the current head. The prior completion-snapshot and effective-tag findings are resolved: the execute sweep uses a frozen dry-run cutoff, confirms completion with a post-processing selection, pins the effective tag across selection, processing, retries, and Continue-As-New, and preserves exact direct-planner approval semantics while validating workflow cohorts against the immutable containing envelope. Pending and due cohorts are selected from one repeatable-read snapshot and merged as row-disjoint aggregates.

Verification:

• Focused unit packages passed locally.
• The focused unit packages also passed 10 consecutive runs.
• Focused go vet passed.
• git diff --check passed.
• ci/circleci: build_and_test is green on this exact head.
